extensions/ql-vscode/src/data-extensions-editor/predicates.ts

```diff
@@ -0,0 +1,136 @@
+import { ExternalApiUsage } from "./external-api-usage";
+import {
+  ModeledMethod,
+  ModeledMethodType,
+  ModeledMethodWithSignature,
+} from "./modeled-method";
+
+export type ExternalApiUsageByType = {
+  externalApiUsage: ExternalApiUsage;
+  modeledMethod: ModeledMethod;
+};
+
+export type ExtensiblePredicateDefinition = {
+  extensiblePredicate: string;
+  generateMethodDefinition: (method: ExternalApiUsageByType) => any[];
+  readModeledMethod: (row: any[]) => ModeledMethodWithSignature;
+
+  supportedKinds?: string[];
+};
+
+function readRowToMethod(row: any[]): string {
+  return `${row[0]}.${row[1]}#${row[3]}${row[4]}`;
+}
+
+export const extensiblePredicateDefinitions: Record<
+  Exclude<ModeledMethodType, "none">,
+  ExtensiblePredicateDefinition
+> = {
+  source: {
+    extensiblePredicate: "sourceModel",
+    // extensible predicate sourceModel(
+    //   string package, string type, boolean subtypes, string name, string signature, string ext,
+    //   string output, string kind, string provenance
+    // );
+    generateMethodDefinition: (method) => [
+      method.externalApiUsage.packageName,
+      method.externalApiUsage.typeName,
+      true,
+      method.externalApiUsage.methodName,
+      method.externalApiUsage.methodParameters,
+      "",
+      method.modeledMethod.output,
+      method.modeledMethod.kind,
+      "manual",
+    ],
+    readModeledMethod: (row) => ({
+      signature: readRowToMethod(row),
+      modeledMethod: {
+        type: "source",
+        input: "",
+        output: row[6],
+        kind: row[7],
+      },
+    }),
+    supportedKinds: ["remote"],
+  },
+  sink: {
+    extensiblePredicate: "sinkModel",
+    // extensible predicate sinkModel(
+    //   string package, string type, boolean subtypes, string name, string signature, string ext,
+    //   string input, string kind, string provenance
+    // );
```

Comment:
Hmmm...this is the extensible predicate signature for static languages only. So far, we support java and csharp. It's likely that go will use the same signature, but not guaranteed.

However, dynamic languages use a different signature. We have a [document that contains a lot of information about data extensions](https://docs.google.com/document/d/14IYCHX8wWuU-HTvJ2gPSdXQKHKYbWCHQKOgn8oLaa80/edit#heading=h.ulpxg3ohlg6g). And it has a link to specific information around javascript.

https://github.com/github/codeql/blob/main/docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst